lib/SIL/SILProfiler.cpp

```diff
@@ -768,26 +795,21 @@ struct CoverageMapping : public ASTWalker {
       return false;
 
     if (auto *AFD = dyn_cast<AbstractFunctionDecl>(D)) {
-      // Don't map a nested function unless it's a constructor.
-      bool continueWalk = Parent.isNull() || isa<ConstructorDecl>(AFD);
-      if (continueWalk) {
+      return visitFunctionDecl(*this, AFD, [&] {
         CounterExpr &funcCounter = assignCounter(AFD->getBody());
 
         if (isa<ConstructorDecl>(AFD))
           addToCounter(ParentNominalType, funcCounter);
-      }
-      return continueWalk;
+      });
     } else if (auto *TLCD = dyn_cast<TopLevelCodeDecl>(D)) {
       assignCounter(TLCD->getBody());
       ImplicitTopLevelBody = TLCD->getBody();
     } else if (auto *NTD = dyn_cast<NominalTypeDecl>(D)) {
-      bool continueWalk = Parent.isNull();
-      if (continueWalk) {
+      return visitNominalTypeDecl(*this, NTD, [&] {
         ParentNominalType = NTD;
```

Comment:
What happens when you have nested types?

Reply:
When a constructor for a type is emitted in SILGen, we lazily create a SILProfiler instance for the nominal type associated with the constructor. The same holds for nested types.

Reply:
Er, that could've been clearer, sorry. We'd create a new ASTWalker for the nested type, where the nested nominal-type-decl would be the root of the walk.